odoo/addons/base/models/res_users.py

```diff
@@ -1618,6 +1619,27 @@ def remove(self):
             return {'type': 'ir.actions.act_window_close'}
         raise AccessError(_("You can not remove API keys unless they're yours or you are a system user"))
 
+    def get_api_key(self, key):
+        index = key[:INDEX_SIZE]
+        self.env.cr.execute('SELECT id, key FROM res_users_apikeys WHERE index = %s', [index])
+        for [id, current_key] in self.env.cr.fetchall():
+            if KEY_CRYPT_CONTEXT.verify(key, current_key):
+                return self.browse(id)
+
```

Comment:
What is this method for, why is it public, and why is it named `get_api_key` when it actually verifies a key? Also keys should not be valid anymore when the user is deactivated (e.g. ignore inactive users in the SELECT by joining res_users)

More generally, I don't think business code should ever access key "records" directly, in the sense of browse()ing them. They're internal data, what the business code needs is simply a way to verify a key and get the corresponding user. A generic api/scope method would be something like (returns UID or user record, either sounds fine):

```py
    def _user_for_key(self, key, scope=None):
        """Verify key and returns the owner user ID"""
        # (...)
        for (user_id, user_key) in self.env.cr.fetchall():
           if KEY_CRYPT_CONTEXT.verify(key, user_key):
               return user_id
        return None
```

Reply:
I agree, though like `_generate_ I would make the two parameters keyword-only and required, with the `scope` first in the argslist (though that's somewhat less important if they're keyword-only). The scope when checking should also probably be non-null for simplicity, with the SQL query updated to check that `scope IS NULL OR scope = %s`.

The RPC layer should also [be modified](https://github.com/odoo/odoo/pull/33928/files#diff-c087b09b8455bbf3fa546a56c6e271bcR1576-R1581) to call this method instead of hand-rolling it, though obviously how to call that scope is non-trivial. I guess `'rpc'` would work, it seems a bit too specific but I can't think of a better one.

The "developer" API key flow should probably still generate keys with an unset scope though, that way they ought work everywhere which makes sense for keys the user generated by hand.

Reply:
So now the code is

```python
    def _check_credentials(self, scope, key):
        index = key[:INDEX_SIZE]
        if scope:
            self.env.cr.execute('SELECT user_id, key FROM res_users_apikeys WHERE index = %s AND (scope IS NULL OR scope = %s)', [index, scope])
        else:
            self.env.cr.execute('SELECT user_id, key FROM res_users_apikeys WHERE index = %s AND scope IS NULL', [index])
        for user_id, current_key in self.env.cr.fetchall():
            if KEY_CRYPT_CONTEXT.verify(key, current_key):
                return user_id
```

- scope as the first parameter and required, as asked by XMO in another comment
- if scope is set, we search for api keys where scope is null too
- however if it is set to None the query is made with only the part `scope is NULL`

So null scoped keys can access everything, however a key with a scope but the wrong one won't be be able to get access. This is intended to be paired with such an auth function:

```python
    @classmethod
    def _auth_method_outlook(cls):
        access_token = request.httprequest.headers.get('access_token')
        if not access_token:
            raise Exception('Access token missing')

        user_id = request.env["res.users.apikeys"]._check_credentials('outlook', access_token)
        if not user_id:
            raise Exception('Access token invalid')

        # take the identity of the API key user
        request.uid = user_id
```